### PR TITLE
MAINT: Remove subclass paths from scalar_value

### PR DESCRIPTION
**Pull request by seberg**
If a user-dtype uses a subclass of an existing NumPy scalar
(which is probably rare or non-existing), it cannot be flexible
or object.
Thus the alignment logic should always work (as it should work
for our scalars as well), and it is unlikely that testing for
subclasses is not just a waste of time.

### CODE REVIEW

File: numpy/core/src/multiarray/scalarapi.c [hunk 1]
```diff
@@ -87,83 +87,12 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
     }
 
     /*
-     * Must be a user-defined type --- check to see which
-     * scalar it inherits from.
+     * Must be a user defined type with an associated (registered) dtype.
+     * Thus, it cannot be flexible (user dtypes cannot be), so we can (and
+     * pretty much have no choice but to) assume the below logic always works.
+     * I.e. this assumes that the logic would also works for most of our types.
```
Comment (rossbar) [inline numpy/core/src/multiarray/scalarapi.c:93]:
```suggestion
     * I.e. this assumes that the logic would also work for most of our types.
```

File: numpy/core/src/multiarray/scalarapi.c [hunk 2]
```diff
@@ -87,83 +87,12 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
...
-        return &PyArrayScalar_VAL(scalar, Bool);
-    }
-    else if (_CHK(Datetime)) {
-        return &PyArrayScalar_VAL(scalar, Datetime);
-    }
-    else if (_CHK(Flexible)) {
-        if (_CHK(String)) {
-            return (void *)PyString_AS_STRING(scalar);
-        }
-        if (_CHK(Unicode)) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalarapi.c:143]: Is it worth throwing an exception for these cases if we no longer support them?

Comment (seberg) [reply numpy/core/src/multiarray/scalarapi.c:143]: Well, this function can't write an error easily. As mentioned in the comment, I do not think it is reachable. Flexible user dtypes are unsupported. So you would need something like a fixed-width string dtype which somehow still subclasses our flexible strings.  On top of that you would need to not use getitem/setitem (which is encouraged) or have some external user use low-level C-API...

But I could check for inheriting from `np.flexible` and return NULL just to provoke a certainly hard crash...

**Comment by mattip** [comment]
@seberg is this meant to be part of the dtype reworking for 1.20 or do we let it slide to a future release?

**Comment by seberg** [comment]
Rebased (though that doesn't decide whether there should be an exception, although we do at dtype creation time).

But no, we can let it slide, this is a simple maintenance.

**Comment by charris** [comment]
close/reopen

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**